packages/NodeTypeResolver/NodeTypeResolver.php

```diff
@@ -350,7 +355,11 @@ private function isObjectTypeOfObjectType(ObjectType $resolvedObjectType, Object
         }
 
         $classReflection = $this->reflectionProvider->getClass($resolvedObjectType->getClassName());
-        if (\trait_exists($requiredObjectType->getClassName())) {
+        if (!isset($this->traitExistsCache[$classReflection->getName()])) {
+            $this->traitExistsCache[$classReflection->getName()] = \trait_exists($requiredObjectType->getClassName());
```

Comment:
calling `trait_exists` or any other runtime autoloading mechanism like `class_exists` in a static analysis context with phpstan is wrong most of the time.

these `*_exists` calls will trigger the real autoloader and might lead to code beeing executed. this in turn can have side effects which can lead to different analysis results based on the order in which files are analyzed.

on the long run, rector should not use `*_exists` php-src functions at all

Reply:
fixed with https://github.com/rectorphp/rector-src/pull/3627